backend/gatekeeper/views.py

```diff
@@ -11,65 +11,79 @@
 from django.urls import reverse
 from django.views.decorators.csrf import csrf_exempt
 from guardian.shortcuts import assign_perm
+from rest_framework.response import Response
+from rest_framework.views import APIView
 
-from gatekeeper.models import Datastream, Thing
-from gatekeeper.utils import parse_sta_url
+from .models import Datastream, Thing
+from .utils import parse_sta_url
 
 
-@csrf_exempt
-def index(request, path):
-    """This is a quick and dirty proof of concept for proxying the SensorThings API queries to the real STA server.
+class Gatekeeper(APIView):
+    sts_self_base_url = None
+    sts_url = None
+    sts_headers = {}
+    sts_arguments = {}
+    sts_content_type = ''
+    dumped_request_data= None
 
-    Additionally when creating Things or Datastreams, this view will create instances of them in the local database.
-    """
-    user = request.user
-    if 'token' in request.GET:
-        user = authenticate(request, token=request.GET.get('token'))
+    def initial(self, request, *args, **kwargs):
+        super().initial(request, *args, **kwargs)
+        path = self.kwargs.pop('path')
 
-    self_base_url = request.build_absolute_uri(reverse('gatekeeper:index', kwargs={
-        'path': ''
-    })).rstrip('/')
+        self.sts_self_base_url = request.build_absolute_uri(reverse('gatekeeper:index', kwargs={
+            'path': ''
+        })).rstrip('/')
 
-    # TODO: validate path
-    url = '{}/{}'.format(settings.GATEKEEPER_STS_BASE_URL, path)
+        # TODO: validate path
+        self.sts_url = '{}/{}'.format(settings.GATEKEEPER_STS_BASE_URL, path)
 
-    request_arguments = {
-        'headers': {
+        self.sts_arguments['headers'] = self.filter_valid_sts_headers(request)
+
+    def filter_valid_sts_headers(self, request):
+        headers = {
             'Content-Type': request.META['CONTENT_TYPE'],
-        },
-    }
+        }
+
+        for header in request.META:
+            if not header.startswith('HTTP_'):
+                continue
+
+            header_name = header[5:]
+            if header_name not in ['ACCEPT', 'EXPECT', 'USER_AGENT']:
+                continue
 
-    for header in request.META:
-        if not header.startswith('HTTP_'):
-            continue
+            headers[header_name] = request.META[header]
+        return headers
 
-        header_name = header[5:]
-        if header_name not in ['ACCEPT', 'EXPECT', 'USER_AGENT']:
-            continue
+    def get(self, request):
+        self.sts_arguments['params'] = {k: v for k, v in request.query_params.items() if k.startswith('$')}
+        return self.handle_request(request)
 
-        request_arguments['headers'][header_name] = request.META[header]
+    def post(self, request):
+        print("Data: \n", self.request.data)
+        self.sts_arguments['json'] = self.request.data
+        return self.handle_request(request)
 
-    # TODO: access control
-    if request.method == 'GET':
-        # TODO: validate GET parameters
```

Comment:
Did you gather the removed TODOs somewhere?

Reply:
Added access control todo to top of class. The validated GET parameters, hmm, I don't know about that one, should be actually do that? Isn't that something that the STS will check, as you said yourself, we should resemble a proxy as much as possibly.